File: typescript/src/features/goose/provider_interface.ts
// Re-implementation of Goose's Abstract LLM Provider Interface

export interface Message {
  role: 'user' | 'assistant' | 'system';
  content: string;
}

export interface ProviderResponse {
  text: string;
  usage?: { promptTokens: number; completionTokens: number };
}

export interface LLMProvider {
  name: string;
  chat(messages: Message[]): Promise<ProviderResponse>;
  streamChat(messages: Message[], onChunk: (chunk: string) => void): Promise<ProviderResponse>;
}

File: go/internal/goose/provider_interface.go
package goose

// Re-implementation of Goose's Abstract LLM Provider Interface

type Role string

const (
	RoleUser      Role = "user"
	RoleAssistant Role = "assistant"
	RoleSystem    Role = "system"
)

type Message struct {
	Role    Role
	Content string
}

type ProviderResponse struct {
	Text             string
	PromptTokens     int
	CompletionTokens int
}

type LLMProvider interface {
	Name() string
	Chat(messages []Message) (ProviderResponse, error)
	StreamChat(messages []Message, onChunk func(chunk string)) (ProviderResponse, error)
}
